src/main/java/io/peacemakr/crypto/impl/crypto/ICryptoImpl.java

```diff
@@ -696,10 +706,70 @@ private AsymmetricKey getSigningKey(SymmetricKeyUseDomain useDomain) {
     AsymmetricKey verificationKey = getOrDownloadPublicKey(aad.senderKeyID);
 
     if (apiKey.equals("")) {
-      Crypto.decryptSymmetric(key, null, cipherText);
+      verificationKey = null;
+    }
+
+    byte[] decryptedMessage = Crypto.decryptSymmetric(key, verificationKey, cipherText);
+    if (decryptedMessage == null)
+      throw new PeacemakrException("Failed to decrypt");
+    return decryptedMessage;
+  }
+
+  @Override
+  public byte[] signOnly(byte[] plainText) throws PeacemakrException {
+    verifyIsBootstrappedAndRegistered();
+
+    // validate the input
+    if (plainText == null)
+      throw new PeacemakrException("Provided plain text array object can not be null");
+    else if (plainText.length == 0) {
+      throw new PeacemakrException("Provided plain text array object can not be empty");
     }
 
-    return Crypto.decryptSymmetric(key, verificationKey, cipherText);
+    // get client's public key id
+    String publicKeyId = this.persister.load(PERSISTER_PREFERRED_KEYID);
+    if (publicKeyId == null || publicKeyId.isEmpty())
+      throw new PeacemakrException("Failed to obtain public key id");
+
+    // Signing key
+    AsymmetricKey signingPrivateKey = getAsymmetricPrivateKey();
+
+    // Construct AAD
+    CiphertextAAD aad = new CiphertextAAD();
+    aad.cryptoKeyID = "";
+    aad.senderKeyID = publicKeyId;
+
+    // ADD bytes
+    Gson gson = new Gson();
+    byte[] addStringBytes = gson.toJson(aad).getBytes(StandardCharsets.UTF_8);
+
+    // Message digest
+    MessageDigest digest = MessageDigest.SHA_256;
+
+    byte[] signedBlob = Crypto.signAsymmetric(signingPrivateKey, plainText, addStringBytes, digest);
+    // Crypto.signAsymmetric might return null if failed to create a signing key
+    if (signedBlob == null)
+      throw new PeacemakrException("Failed to sign plain text");
+    return signedBlob;
+  }
+
+  @Override
+  public byte[] verifyOnly(byte[] signedBlob) throws PeacemakrException {
+    // validate the input
+    if (signedBlob == null)
+      throw new PeacemakrException("Provided signed blob array object can not be null");
+    else if (signedBlob.length == 0) {
+      throw new PeacemakrException("Provided signed blob array object can not be empty");
+    }
+
+    // Private key
+    AsymmetricKey signingPrivateKey = getAsymmetricPrivateKey();
+
+    byte[] verifiedBlob = Crypto.verifyAsymmetric(signingPrivateKey, signedBlob);
+    // null is returned if failed to verify
+    if (verifiedBlob == null)
+      throw new PeacemakrException("Failed to verify signed blob");
```

Comment:
"failed to verify" vs "verification failed" might be important distinctions at some point?